File: src/com/aopphp/go/injector/PointcutQueryLanguageInjector.java
package com.aopphp.go.injector;

import com.aopphp.go.PointcutQueryLanguage;
import com.aopphp.go.pattern.CodePattern;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.InjectedLanguagePlaces;
import com.intellij.psi.LanguageInjector;
import com.intellij.psi.PsiLanguageInjectionHost;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import org.jetbrains.annotations.NotNull;

/**
 * PointcutQuery injector is responsible to inject Go! AOP Pointcut syntax into the StringLiteral expressions
 */
public class PointcutQueryLanguageInjector implements LanguageInjector {

    private static final String GO_AOP_ANNOTATION = "\\Go\\Lang\\Annotation";

    @Override
    public void getLanguagesToInject(
            @NotNull PsiLanguageInjectionHost host,
            @NotNull InjectedLanguagePlaces injectionPlacesRegistrar)
    {
        // we accept only PHP literal expressions, such as docBlocks and string for PointcutBuilder->method('<expr>')
        if (!(host instanceof StringLiteralExpression)) {
            return;
        }

        StringLiteralExpression literalExpression = (StringLiteralExpression) host;

        boolean enableInjection = CodePattern.isInsideAnnotation(literalExpression, GO_AOP_ANNOTATION);
        enableInjection |= CodePattern.isInsidePointcutBuilderMethod(literalExpression);

        if (enableInjection) {
            TextRange rangeInsideHost = new TextRange(1, Math.max(literalExpression.getTextLength()-1, 1));
            injectionPlacesRegistrar.addPlace(PointcutQueryLanguage.INSTANCE, rangeInsideHost, null, null);
        }
    }
}

File: src/main/kotlin/com/aopphp/go/injector/PointcutQueryLanguageInjector.kt
package com.aopphp.go.injector

import com.aopphp.go.PointcutQueryLanguage
import com.aopphp.go.pattern.CodePattern
import com.intellij.lang.injection.MultiHostInjector
import com.intellij.lang.injection.MultiHostRegistrar
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiLanguageInjectionHost
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import kotlin.math.max

/**
 * Injects the Go! AOP Pointcut query language into PHP string literals that are:
 *  - arguments of #[\Go\Lang\Attribute\*] PHP 8 attributes
 *  - arguments of PointcutBuilder->method() calls
 *
 * Uses MultiHostInjector so both single-quoted and double-quoted strings are covered.
 */
class PointcutQueryLanguageInjector : MultiHostInjector {

    override fun getLanguagesToInject(registrar: MultiHostRegistrar, context: PsiElement) {
        if (context !is StringLiteralExpression) return
        val inject = CodePattern.isInsidePhpAttribute(context, GO_AOP_ATTRIBUTE_PREFIX)
                  || CodePattern.isInsidePointcutBuilderMethod(context)
        if (!inject) return
        val range = TextRange(1, max(context.textLength - 1, 1))
        registrar.startInjecting(PointcutQueryLanguage)
            .addPlace(null, null, context as PsiLanguageInjectionHost, range)
            .doneInjecting()
    }

    override fun elementsToInjectIn(): List<Class<out PsiElement>> =
        listOf(StringLiteralExpression::class.java)

    companion object {
        private const val GO_AOP_ATTRIBUTE_PREFIX = "\\Go\\Lang\\Attribute"
    }
}
